runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java

```diff
@@ -325,74 +357,63 @@ private FlinkAutoBalancedShardKeyShardingFunction(
-      if (cache.size() < shardNumber) {
-        for (int i = 0; i < shardNumber; i++) {
-          generateInternal(new CacheKey(destinationKey, i));
-        }
+      if (cache.getIfPresent(destinationKey) == null) {
+        cache.put(destinationKey, generateShardedKeys(destinationKey, shardCount));
       }
 
-      return generateInternal(new CacheKey(destinationKey, shardNumber));
+      return cache.getIfPresent(destinationKey).get(shardNumber);
     }
 
-    private ShardedKey<Integer> generateInternal(CacheKey key) {
-
-      ShardedKey<Integer> result = cache.get(key);
-      if (result != null) {
-        return result;
-      }
-
-      int salt = -1;
-      while (true) {
-        salt++;
-        ShardedKey<Integer> shk = ShardedKey.of(Objects.hash(key.key, salt), key.shard);
-        int targetPartition = key.shard % parallelism;
-
-        // create effective key in the same way Beam/Flink will do so we can see if it gets
-        // allocated to the partition we want
-        ByteBuffer effectiveKey;
-        try {
-          byte[] bytes = CoderUtils.encodeToByteArray(ShardedKeyCoder.of(VarIntCoder.of()), shk);
-          effectiveKey = ByteBuffer.wrap(bytes);
-        } catch (CoderException e) {
-          throw new RuntimeException(e);
-        }
-
-        int partition =
-            KeyGroupRangeAssignment.assignKeyToParallelOperator(
-                effectiveKey, maxParallelism, parallelism);
-
-        if (partition == targetPartition && !usedSalts.contains(salt)) {
-          usedSalts.add(salt);
-          cache.put(key, shk);
-          return shk;
+    private Map<Integer, ShardedKey<Integer>> generateShardedKeys(int key, int shardCount) {
+
+      Set<Integer> usedSalts = new HashSet<>();
+      Map<Integer, ShardedKey<Integer>> shardedKeys = new HashMap<>();
+
+      for (int shard = 0; shard < shardCount; shard++) {
+
+        int salt = -1;
+        while (true) {
+          if (salt++ == Integer.MAX_VALUE) {
+            throw new RuntimeException(
+                "Failed to find sharded key in [ " + Integer.MAX_VALUE + " ] iterations");
+          }
+          ShardedKey<Integer> shk = ShardedKey.of(Objects.hash(key, salt), shard);
+          int targetPartition = shard % parallelism;
+
+          // create effective key in the same way Beam/Flink will do so we can see if it gets
+          // allocated to the partition we want
+          ByteBuffer effectiveKey;
+          try {
+            byte[] bytes = CoderUtils.encodeToByteArray(ShardedKeyCoder.of(VarIntCoder.of()), shk);
+            effectiveKey = ByteBuffer.wrap(bytes);
+          } catch (CoderException e) {
+            throw new RuntimeException(e);
+          }
+
+          int partition =
+              KeyGroupRangeAssignment.assignKeyToParallelOperator(
+                  effectiveKey, maxParallelism, parallelism);
+
+          if (partition == targetPartition && !usedSalts.contains(salt)) {
```

Comment:
Why do we need `!usedSalts.contains(salt)`?

Reply:
shard key identity is defined by destination (it's hash), shardNumber and salt. Salt is integer from we are searching for, to make sure the "real key" (destination, shardNumber) is landed to correct worker. Salt is restarted for each shard key to zero. 

I guess I was paranoid in the beginning to avoid key collisions. But right now, I do not see a reason why not to drop it.